aten/src/ATen/native/cuda/Indexing.cu

```diff
@@ -1243,5 +1246,213 @@ Tensor & masked_fill__cuda(Tensor& self, const Tensor & mask, const Tensor & val
+  if (dim < sparse_dim) {
+    // short-circuit if index is empty
+    if (!index_len) {
+      auto res_indices = indices.index_select(1, index);
+      res_indices[dim] = index;
+      const auto res_values = values.index_select(0, index);
+
+      return at::_sparse_coo_tensor_with_dims_and_tensors(
+          sparse_dim, dense_dim, res_sizes, res_indices, res_values, self.options());
+    }
+
+    const auto nneg_index = [&index, size]() -> Tensor {
+      auto nneg_index = at::empty_like(index, at::MemoryFormat::Contiguous);
+
+      auto iter = TensorIteratorConfig()
+        .add_output(nneg_index)
+        .add_input(index)
+        .build();
+
+      AT_DISPATCH_INDEX_TYPES(index.scalar_type(), "index_select_sparse_cuda", [&]() {
+          gpu_kernel(iter, [size] GPU_LAMBDA (index_t idx) -> index_t {
+              CUDA_KERNEL_ASSERT(idx >= -size && idx < size
+                  && "index_select(): index out of bounds");
+              return idx < 0 ? idx + size : idx;
+          });
+      });
+      return nneg_index;
+    }();
+    const auto dim_indices = indices[dim].contiguous();
+
+    const auto idx_nneg_index = at::arange(index_len, nneg_index.options());
+    const auto idx_dim_indices = at::arange(nnz, dim_indices.options());
+
+    const auto intrsc_counts_nneg_index = [&]() -> Tensor {
+      auto intrsc_counts_nneg_index = at::zeros_like(nneg_index);
+
+      const auto rows_over_dim_indices = (nnz <= index_len);
+      // Need to have output as TensorIterator does not allow having void lambdas.
+      auto dummy_output = at::empty({1, 1}, dim_indices.options())
+        .expand(
+            rows_over_dim_indices
+            ? IntArrayRef({nnz, index_len})
+            : IntArrayRef({index_len, nnz})
+        );
+      auto iter = TensorIteratorConfig()
+        .add_output(dummy_output)
+        // All iterations map to a single element in dummy_output by design,
+        // hence removed output memory overlap check.
+        .set_check_mem_overlap(false)
+        .add_owned_input(dim_indices.unsqueeze(rows_over_dim_indices ? -1: -2))
+        .add_owned_input(nneg_index.unsqueeze(rows_over_dim_indices ? -2: -1))
+        .add_owned_input(idx_nneg_index.unsqueeze(rows_over_dim_indices ? -2: -1))
+        .build();
+
+      AT_DISPATCH_INDEX_TYPES(nneg_index.scalar_type(), "index_select_sparse_cuda", [&]() {
+          index_t* ptr_intrsc_counts_nneg_index = intrsc_counts_nneg_index.data_ptr<index_t>();
+          gpu_kernel(
+              iter,
+              [ptr_intrsc_counts_nneg_index, index_len] GPU_LAMBDA (
+                index_t dim_val, index_t idx_val, // dim_indices[j], j
+                index_t idx_idx // i
+              ) -> index_t {
+                if (dim_val == idx_val) {
+                  applyFastAtomicIndexInc(
+                    ptr_intrsc_counts_nneg_index,
+                    idx_idx,
+                    static_cast<index_t>(index_len)
+                  );
```

Comment:
@cpuhrsch , the regressed case `n=10, nnz=10000, index_len=1000` really fails here. Since `n=10` we have on average about 100 threads trying to write into the same memory, and synchronizations on CUDA are pretty slow... Therefore I have this question, how likely it is that we have such a case for GPUs? Do we merge this and wait on improvement, or shall we just implement these improvements in this PR?

Reply:
@nikitaved - I'm not sure how likely we're to hit this case, because we don't have integrations into reference models for this. Can't we go back to our CPU-fallback baseline for this case if it's easy to detect and optimize once we've integrated this function into a model and this case shows up?

Reply:
Not relevant anymore, with the newest update we are golden.